classes/components/forms/publication/ContributorForm.php

```diff
@@ -0,0 +1,22 @@
+<?php
+/**
+ * @file classes/components/form/publication/ContributorForm.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
+ * Copyright (c) 2000-2022 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class ContributorForm
+ * @ingroup classes_controllers_form
+ *
+ * @brief A preset form for adding and editing a contributor for a publication.
+ */
+
+namespace APP\components\forms\publication;
+
+use PKP\components\forms\publication\PKPContributorForm;
```

Comment:
What do you think about using the Laravel service container for things like this? Since we're now namespaced we no longer need the `PKP` suffixes on classes (except for backwards compatibility). We can let the app configure the service container with the appropriate classname when we need to specify it, and have pkp-lib just specify `ContributorForm`. Particularly when there are empty subclasses, I find myself wishing for a better way...

Reply:
The thing I don't like about using `app(ContributorForm::class)` or `App::make(ContributorForm::class)` is that the return is not type hinted. So it has to be manually annotated:

```php
/** @var ContributorForm $form */
$form = app(ContributorForm::class);
```

The other problem is that if dependencies can't be automatically injected -- and they usually can't be for forms -- then we have to read the code to know what to pass to the service container, rather than having the editor tell us what the constructor expects:

```php
$form = app(ContributorForm::class, [
    'action' => $url,
    'locales' => $locales,
    'submission' => $submission,
    'context' => $context,
]);
```

This becomes especially problematic if the constructors differ between apps. And using different constructor function signatures is a common way to extend a base class in an app.

I think the service container is only good for when we want to do automatic dependency injection. For example, injecting the request, file service, schema service or other singletons into a `Handler.

Reply:
> I find myself wishing for a better way...

Increasingly, I find that I am writing a helper function in a `PageHandler` that gets a form class, because this can be clearly type-hinted and easily overridden in an app:

```php
namespace PKP\pages;

use PKP\components\SectionForm;

class SectionHandler extends Handler
{
    protected function getSectionForm(Context $context): SectionForm
    {
        return new SectionForm($context);
    }
}
```

```php
namespace APP\pages\section;

use APP\components\SectionForm;

class SectionHandler extends PKP\pages\SectionHandler
{
    protected function getSectionForm(Context $context): SectionForm
    {
        return new SectionForm($context);
    }
}
```

That's a lot of boilerplate. But the base `Handler` is fully decoupled from `APP\components\SectionForm` and everything is nicely type-hinted. If the base `Handler` needs a function that's dramatically different between apps, then `PKP\components\SectionForm` can add an abstract function for it.

Reply:
Less boilerplate than a pointless empty subclass. I think this is a good approach.